main.cpp

```diff
@@ -2027,15 +2104,73 @@ int main(int argc, char **argv) {
                         rc = ERROR_NO_DEVICE;
+                    } else if (supported == cmd::device_support::one) {
+                        if (devices[dr_vidpid_bootrom_ok].size() > 1 ||
+                            (devices[dr_vidpid_bootrom_ok].empty() && devices[dr_vidpid_stdio_usb].size() > 1)) {
+                            fail(ERROR_NOT_POSSIBLE, "Command requires a single RP2040 device to be targeted.");
+                        }
+                        if (!devices[dr_vidpid_bootrom_ok].empty()) {
+                            settings.force = false; // we have a device, so we're not forcing
+                        }
                     }
+                    break;
                 default:
                     break;
             }
-            if (!rc) selected_cmd->execute(devices);
+            if (!rc) {
+                if (settings.force && ctx) { // actually ctx should never be null if force is set, but still
+                    if (devices[dr_vidpid_stdio_usb].size() != 1) {
+                        fail(ERROR_NOT_POSSIBLE,
+                             "Forced command requires a single rebootable RP2040 device to be targeted.");
+                    }
+                    if (selected_cmd->force_requires_pre_reboot()) {
+                        // we reboot into BOOTSEL mode and disable MSC interface (the 1 here)
+                        auto &to_reboot = devices[dr_vidpid_stdio_usb][0].first;
+                        reboot_device(to_reboot, true, 1);
+                        std::cout
+                                << "The device was rebooted as requested into BOOTSEL mode so the command can be executed.\n\n";
+                        for (const auto &handle : to_close) {
+                            libusb_close(handle);
+                        }
+                        libusb_free_device_list(devs, 1);
+                        devs = nullptr;
+                        to_close.clear();
+                        devices.clear();
+                        sleep_ms(500);
+
+                        // we now clear settings.force, because we expect the device to have rebooted and be available.
+                        // we also clear any filters, because the device may have moved, so the only way we can find it
+                        // again is to assume it is the only now visible device.
+                        settings.force = false;
+                        settings.address = -1;
+                        settings.bus = -1;
+                        continue;
+                    }
+                }
+                selected_cmd->execute(devices);
+                if (tries) {
+                    if (settings.no_reboot_if_forced) {
+                        printf("The device has been left accessible, but without the drive mounted; use 'picotool reboot' to reboot into regular BOOTSEL mode or application mode.");
```

Comment:
Why does this use `printf` when `cout` is used elsewhere? Also, is the trailing `\n` missing?

Reply:
Yep, this is definitely missing a `\n`...
![image](https://user-images.githubusercontent.com/476186/121365735-d0957100-c930-11eb-91bc-ce9ea15b9716.png)